consul/server_manager/server_manager.go

```diff
@@ -0,0 +1,340 @@
+	}
+}
+
+// NumServers takes out an internal "read lock" and returns the number of
+// servers.  numServers includes both healthy and unhealthy servers.
+func (sm *ServerManager) NumServers() (numServers int) {
+	serverCfg := sm.getServerConfig()
+	numServers = len(serverCfg.servers)
+	return numServers
+}
+
+// RebalanceServers takes out an internal write lock and shuffles the list of
+// servers on this agent.  This allows for a redistribution of work across
+// consul servers and provides a guarantee that the order of the server list
+// isn't related to the age at which the node was added to the cluster.
+// Elsewhere we rely on the position in the server list as a hint regarding
+// the stability of a server relative to its position in the server list.
+// Servers at or near the front of the list are more stable than servers near
+// the end of the list.
+func (sm *ServerManager) RebalanceServers() {
+	sm.serverConfigLock.Lock()
+	defer sm.serverConfigLock.Unlock()
+	serverCfg := sm.getServerConfig()
+
+	newServers := make([]*server_details.ServerDetails, len(serverCfg.servers))
+	copy(newServers, serverCfg.servers)
+
+	// Shuffle the server list on server join.  Servers are selected from
+	// the head of the list and are moved to the end of the list on
+	// failure.
+	for i := len(serverCfg.servers) - 1; i > 0; i-- {
+		j := rand.Int31n(int32(i + 1))
+		newServers[i], newServers[j] = newServers[j], newServers[i]
+	}
+	serverCfg.servers = newServers
+
+	sm.saveServerConfig(serverCfg)
+}
+
+// RemoveServer takes out an internal write lock and removes a server from
+// the server list.
+func (sm *ServerManager) RemoveServer(server *server_details.ServerDetails) {
+	sm.serverConfigLock.Lock()
+	defer sm.serverConfigLock.Unlock()
+	serverCfg := sm.getServerConfig()
+
+	// Remove the server if known
+	n := len(serverCfg.servers)
+	for i := 0; i < n; i++ {
+		if serverCfg.servers[i].Name == server.Name {
+			newServers := make([]*server_details.ServerDetails, 0, len(serverCfg.servers)-1)
+			newServers = append(newServers, serverCfg.servers[:i]...)
+			newServers = append(newServers, serverCfg.servers[i+1:]...)
+			serverCfg.servers = newServers
+
+			sm.saveServerConfig(serverCfg)
+			return
+		}
+	}
+}
+
+// refreshServerRebalanceTimer is only called once the rebalanceTimer
+// expires.  Historically this was an expensive routine and is intended to be
+// run in isolation in a dedicated, non-concurrent task.
+func (sm *ServerManager) refreshServerRebalanceTimer(timer *time.Timer) time.Duration {
+	serverCfg := sm.getServerConfig()
+	numConsulServers := len(serverCfg.servers)
+	// Limit this connection's life based on the size (and health) of the
+	// cluster.  Never rebalance a connection more frequently than
+	// connReuseLowWatermarkDuration, and make sure we never exceed
+	// clusterWideRebalanceConnsPerSec operations/s across numLANMembers.
+	clusterWideRebalanceConnsPerSec := float64(numConsulServers * newRebalanceConnsPerSecPerServer)
```

Comment:
If there are no servers than this rate will go to zero, which will result in a divide by zero in `RateScaledInterval` - need a guard here against that (or over there, have it go to the minimum interval for the zero rate).